pkg/controllers/nodeclass/capacityreservation.go

```diff
@@ -0,0 +1,133 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclass
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/capacityreservation"
+)
+
+const capacityReservationPollPeriod = time.Minute
+
+type CapacityReservation struct {
+	provider capacityreservation.Provider
+	clk      clock.Clock
+	cm       *pretty.ChangeMonitor
+}
+
+func NewCapacityReservationReconciler(clk clock.Clock, provider capacityreservation.Provider) *CapacityReservation {
+	return &CapacityReservation{
+		provider: provider,
+		clk:      clk,
+		cm:       pretty.NewChangeMonitor(),
+	}
+}
+
+func (c *CapacityReservation) Reconcile(ctx context.Context, nc *v1.EC2NodeClass) (reconcile.Result, error) {
+	reservations, err := c.provider.List(ctx, nc.Spec.CapacityReservationSelectorTerms...)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting capacity reservations, %w", err)
+	}
+	if len(reservations) == 0 {
+		nc.StatusConditions().SetTrue(v1.ConditionTypeCapacityReservationsReady)
+		return reconcile.Result{RequeueAfter: capacityReservationPollPeriod}, nil
+	}
+
+	if ids := lo.Map(reservations, func(r *ec2types.CapacityReservation, _ int) string {
+		return *r.CapacityReservationId
+	}); c.cm.HasChanged(nc.Name, ids) {
+		log.FromContext(ctx).V(1).WithValues("ids", ids).Info("discovered capacity reservations")
+	}
+	sort.Slice(reservations, func(i, j int) bool {
+		return *reservations[i].CapacityReservationId < *reservations[j].CapacityReservationId
+	})
+	errors := []error{}
+	nc.Status.CapacityReservations = []v1.CapacityReservation{}
+	for _, r := range reservations {
+		reservation, err := capacityReservationFromEC2(r)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		nc.Status.CapacityReservations = append(nc.Status.CapacityReservations, reservation)
+	}
+	if len(errors) != 0 {
+		log.FromContext(ctx).Error(multierr.Combine(errors...), "failed to update status with %d of %d capacity reservations", len(errors), len(reservations))
+	}
+	nc.StatusConditions().SetTrue(v1.ConditionTypeCapacityReservationsReady)
+	return reconcile.Result{RequeueAfter: c.requeueAfter(reservations...)}, nil
+}
+
+func capacityReservationFromEC2(cr *ec2types.CapacityReservation) (v1.CapacityReservation, error) {
+	if !lo.Contains([]ec2types.InstanceMatchCriteria{
```

Comment:
Is this to guard our validation logic from EC2 supporting additional instance match criteria in the future? If so, can we say so in a comment?

Reply:
Yeah, I can add a comment